### PR TITLE
Initialize 'Case' action for Salesforce

### PR DESCRIPTION
**Pull request by LLuque-twilio**
_A summary of your pull request, including the what change you're making and why._

This PR addresses [STRATCONN 1037](https://segment.atlassian.net/browse/STRATCONN-1037) JIRA ticket.

It introduces the 'Case' action to the Salesforce Destination. It includes the three functionalities create, update, and upsert for Cases.

## Testing

_Include any additional information about the testing you have completed to
ensure your changes behave as expected. For a speedy review, please check
any of the tasks you completed below during your testing._

- [x] Added [unit tests](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing) for new functionality
Screen Shot 2022-02-15 at 11 45 11 AM

- [x] Tested end-to-end using the [local server](https://github.com/segmentio/action-destinations/blob/main/docs/testing.md#local-end-to-end-testing)

- [x] [Segmenters] Tested in the staging environment
Screen Shot 2022-02-10 at 3 59 36 PM
Screen Shot 2022-02-10 at 3 59 45 PM

### CODE REVIEW

File: packages/destination-actions/src/destinations/salesforce/sf-object-to-shape.ts [hunk 1]
```diff
@@ -14,7 +14,16 @@ const LeadShape = (payload: GenericPayload): LeadBaseShapeType => {
...
 }
 
-const objectToShape = new Map([['lead', LeadShape]])
+const CaseShape = (payload: GenericPayload): CaseBaseShapeType => {
+  return {
+    Description: payload.description
+  }
+}
+
+const objectToShape = new Map<string, (payload: GenericPayload) => any>([
```
Comment (nick-Ag) [inline packages/destination-actions/src/destinations/salesforce/sf-object-to-shape.ts:23]:
```suggestion
const objectToShape = new Map<string, (payload: GenericPayload) => LeadBaseShapeType | CaseBaseShapeType>([
```
Let's do this instead, it'll get clunky as we add actions but that's fine

File: packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+          traits: {
+            description: 'Test one'
+          },
+          description: {
+            '@path': '$.traits.description'
+          }
+        }
+      })
+
+      console.log(responses)
```
Comment (jinapark202) [inline packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts:251]: You forgot a console.log here

**Review by nick-Ag — APPROVED** [review]
Looks great :shipit:

File: packages/destination-actions/src/destinations/salesforce/index.ts [hunk 3]
```diff
@@ -1,5 +1,7 @@
 import type { DestinationDefinition } from '@segment/actions-core'
 import type { Settings } from './generated-types'
+// This has to be 'cases' because 'case' is a Javascript reserved word
+import cases from './case'
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/index.ts:4]: Not a blocker. In the future, we could consider just using `caseObject`, `leadObject`, etc., but `cases` is ok, too!

File: packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+  describe('Case', () => {
+    it('should create a case record', async () => {
+      nock(`${settings.instanceUrl}/services/data/${API_VERSION}/sobjects`).post('/Case').reply(201, {})
+
+      const event = createTestEvent({
+        event: 'Identify',
+        traits: {
+          description: 'This is test description'
+        }
+      })
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts:26]: I see that these test payloads were based on prior art, so our bad here! I would recommend some changes to ensure that the events conform to our [spec](https://segment.com/docs/connections/spec/):

```suggestion
      const event = createTestEvent({
        type: 'track',
        event: 'Create new case',
        properties: {
          description: 'This is test description'
        }
      })
```

File: packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+
+    it('should create a case record with custom fields', async () => {
+      nock(`${settings.instanceUrl}/services/data/${API_VERSION}/sobjects`).post('/Case').reply(201, {})
+
+      const event = createTestEvent({
+        event: 'Identify',
+        traits: {
+          description: 'This is test description'
+        }
+      })
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts:70]:
```suggestion
      const event = createTestEvent({
        type: 'track',
        event: 'Create new case',
        properties: {
          description: 'This is test description'
        }
      })
```

File: packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+      )
+    })
+
+    it('should update a case record', async () => {
+      const event = createTestEvent({
+        event: 'Identify',
+        traits: {
+          description: 'Test two'
+        }
+      })
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts:119]:
```suggestion
    it('should update a case record', async () => {
      const event = createTestEvent({
        type: 'track',
        event: 'Update case',
        properties: {
          description: 'Test two'
        }
      })
```

File: packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+      expect(responses[1].options.body).toMatchInlineSnapshot(`"{\\"Description\\":\\"Test two\\"}"`)
+    })
+
+    it('should upsert a non-existent record', async () => {
+      const event = createTestEvent({
+        event: 'Identify',
+        traits: {
+          description: 'Test two'
+        }
+      })
```
Comment (brennan) [inline packages/destination-actions/src/destinations/salesforce/__tests__/case.test.ts:225]:
```suggestion
      const event = createTestEvent({
        type: 'track',
        event: 'Upsert case',
        properties: {
          description: 'Test two'
        }
      })
```

**Pull request merged**